SW_Flow_lib.c

```diff
@@ -2158,45 +2193,51 @@ void soil_temperature(double airTemp, double pet, double aet, double biomass,
 	if (debug) {
 		swprintf("\nSoil temperature profile values:");
 		for (i = 0; i <= nRgr + 1; i++) {
-			swprintf("\nk %d oldsTempR %f sTempR %f depth %f",
-				i, st->oldsTempR[i], sTempR[i], (i * deltaX)); // *(oldsTempR + i) is equivalent to writing oldsTempR[i]
+			swprintf("\nk %d oldavgLyrTempR %f avgLyrTempR %f depth %f",
+				i, st->oldavgLyrTempR[i], avgLyrTempR[i], (i * deltaX)); // *(oldavgLyrTempR + i) is equivalent to writing oldavgLyrTempR[i]
 		}
 		swprintf("\n");
 	}
 	#endif
 
-	// convert soil temperature of soil temperature profile 'sTempR' to soil profile layers 'sTemp'
-	lyrTemp_to_lyrSoil_temperature(st->tlyrs_by_slyrs, nRgr, st->depthsR, sTempR, nlyrs,
-		st->depths, width, sTemp);
+	// convert soil temperature of soil temperature profile 'avgLyrTempR' to soil profile layers 'avgLyrTemp'
+	lyrTemp_to_lyrSoil_temperature(st->tlyrs_by_slyrs, nRgr, st->depthsR, avgLyrTempR, nlyrs,
+		st->depths, width, avgLyrTemp, st->temperatureRangeR, v->temperatureRange);
 
 	// Calculate fusion pools based on soil profile layers, soil freezing/thawing, and if freezing/thawing not completed during one day, then adjust soil temperature
-	sFadjusted_sTemp = adjust_Tsoil_by_freezing_and_thawing(oldsTemp, sTemp, shParam,
+	sFadjusted_avgLyrTemp = adjust_Tsoil_by_freezing_and_thawing(oldavgLyrTemp, avgLyrTemp, shParam,
 		nlyrs, vwc, bDensity);
 
-	// update sTempR if sTemp were changed due to soil freezing/thawing
-	if (sFadjusted_sTemp) {
-		lyrSoil_to_lyrTemp_temperature(nlyrs, st->depths, sTemp, sTconst, nRgr, st->depthsR,
-			theMaxDepth, sTempR);
+	// update avgLyrTempR if avgLyrTemp were changed due to soil freezing/thawing
+	if (sFadjusted_avgLyrTemp) {
+		lyrSoil_to_lyrTemp_temperature(nlyrs, st->depths, avgLyrTemp, sTconst, nRgr, st->depthsR,
+			theMaxDepth, avgLyrTempR);
 	}
 
 	// determine frozen/unfrozen status of soil layers
-	set_frozen_unfrozen(nlyrs, sTemp, swc, swc_sat, width);
+	set_frozen_unfrozen(nlyrs, avgLyrTemp, swc, swc_sat, width);
+    
+    // Add/subtract the interpolated range to/from average layer temperature
+    for(i = 0; i < nlyrs; i++) {
+        v->maxLyrTemperature[i] = avgLyrTemp[i] + (v->temperatureRange[i] / 2);

```

Comment:
Could we please make `maxLyrTemperature[]` and `minLyrTemperature[]` arguments to `soil_temperature()` -- to avoid global variables